File: CpuThrottlingTest/StabilityChart.swift
import SwiftUI

struct StabilityChart: View {
    let points: [StabilityPoint]
    let events: [ThermalEvent]
    
    @State private var selectedPoint: StabilityPoint? = nil
    @State private var touchLocationX: CGFloat? = nil
    
    // Gradient definitions for the line and filled area
    private var chartGradient: LinearGradient {
        LinearGradient(
            stops: [
                .init(color: Color(red: 0.2, green: 0.8, blue: 0.4), location: 0.0),  // 100% - Green
                .init(color: Color(red: 0.95, green: 0.7, blue: 0.1), location: 0.25), // 75% - Amber
                .init(color: Color(red: 0.9, green: 0.35, blue: 0.1), location: 0.5),  // 50% - Deep Orange
                .init(color: Color(red: 0.85, green: 0.2, blue: 0.2), location: 1.0)   // 0% - Red
            ],
            startPoint: .top,
            endPoint: .bottom
        )
    }
    
    private var fillGradient: LinearGradient {
        LinearGradient(
            stops: [
                .init(color: Color(red: 0.2, green: 0.8, blue: 0.4).opacity(0.15), location: 0.0),
                .init(color: Color(red: 0.95, green: 0.7, blue: 0.1).opacity(0.1), location: 0.25),
                .init(color: Color(red: 0.9, green: 0.35, blue: 0.1).opacity(0.05), location: 0.5),
                .init(color: Color.clear, location: 1.0)
            ],
            startPoint: .top,
            endPoint: .bottom
        )
    }
    
    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            // Chart Title & Touch Info Header (Disappears when touch is released)
            HStack {
                Image(systemName: "chart.xyaxis.line")
                    .font(.system(size: 14, weight: .semibold))
                    .foregroundColor(.secondary)
                Text("UNIFIED STABILITY CHART")
                    .font(.system(size: 12, weight: .bold, design: .monospaced))
                    .foregroundColor(.secondary)
                Spacer()
                
                if touchLocationX != nil, let pt = selectedPoint {
                    HStack(spacing: 6) {
                        Text(formatTime(pt.time))
                            .font(.system(size: 10, weight: .bold, design: .monospaced))
                            .foregroundColor(Color(red: 0.0, green: 0.9, blue: 1.0))
                        Text("SCORE: \(String(format: "%.1f", pt.score))%")
                            .font(.system(size: 10, weight: .black, design: .monospaced))
                            .foregroundColor(pt.score >= 90 ? Color.green : (pt.score >= 75 ? Color.orange : Color.red))
                    }
                    .padding(.horizontal, 8)
                    .padding(.vertical, 3)
                    .background(Color(red: 0.0, green: 0.9, blue: 1.0).opacity(0.15))
                    .cornerRadius(8)
                    .overlay(
                        RoundedRectangle(cornerRadius: 8)
                            .stroke(Color(red: 0.0, green: 0.9, blue: 1.0).opacity(0.3), lineWidth: 1)
                    )
                }
            }
            .padding(.horizontal, 4)
            
            GeometryReader { geometry in
                let width = geometry.size.width
                let height = geometry.size.height
                
                ZStack(alignment: .topLeading) {
                    // 1. Minimal Grid Planes
                    gridLines(width: width, height: height)
                    
                    // 2. Line Chart & Area Fill
                    if points.count > 0 {
                        // Area fill
                        chartPath(in: geometry.size, isFilled: true)
                            .fill(fillGradient)
                        
                        // Bezier path line
                        chartPath(in: geometry.size, isFilled: false)
                            .stroke(
                                chartGradient,
                                style: StrokeStyle(lineWidth: 3.5, lineCap: .round, lineJoin: .round)
                            )
                            .shadow(color: Color.black.opacity(0.3), radius: 3, x: 0, y: 2)
                    }
                    
                    // 3. Thermal Event Annotations
                    ForEach(events) { event in
                        if let xPos = getXPosition(for: event.time, width: width) {
                            Path { path in
                                path.move(to: CGPoint(x: xPos, y: 15))
                                path.addLine(to: CGPoint(x: xPos, y: height))
                            }
                            .stroke(event.color.opacity(0.5), style: StrokeStyle(lineWidth: 1, dash: [4, 4]))
                        }
                    }
                    
                    // 4. Interactive Touch Selector Line & Node (Only visible while touching)
                    if let tx = touchLocationX, points.count > 0 {
                        let nearest = findNearestPoint(x: tx, width: width)
                        if let targetPt = nearest, let targetX = getXPosition(for: targetPt.time, width: width) {
                            let targetY = height - (CGFloat(targetPt.score / 100.0) * height)
                            
                            // Dotted selector line
                            Path { path in
                                path.move(to: CGPoint(x: targetX, y: 0))
                                path.addLine(to: CGPoint(x: targetX, y: height))
                            }
                            .stroke(Color(red: 0.0, green: 0.9, blue: 1.0).opacity(0.7), style: StrokeStyle(lineWidth: 1.5, dash: [6, 4]))
                            
                            // Outer aura node
                            Circle()
                                .fill(Color(red: 0.0, green: 0.9, blue: 1.0).opacity(0.25))
                                .frame(width: 20, height: 20)
                                .position(x: targetX, y: targetY)
                            
                            // Inner node
                            Circle()
                                .fill(Color(red: 0.0, green: 0.9, blue: 1.0))
                                .frame(width: 8, height: 8)
                                .position(x: targetX, y: targetY)
                        }
                    }
                }
                .contentShape(Rectangle())
                .gesture(
                    DragGesture(minimumDistance: 0)
                        .onChanged { value in
                            let locX = min(width, max(0, value.location.x))
                            touchLocationX = locX
                            selectedPoint = findNearestPoint(x: locX, width: width)
                        }
                        .onEnded { _ in
                            touchLocationX = nil
                            selectedPoint = nil
                        }
                )
            }
            .frame(height: 180)
            
            // X-Axis Timeline Indicator
            HStack {
                Text("0m:00s")
                Spacer()
                if let lastPoint = points.last {
                    Text(formatTime(lastPoint.time))
                } else {
                    Text("0m:00s")
                }
            }
            .font(.system(size: 10, weight: .semibold, design: .monospaced))
            .foregroundColor(.secondary)
            .padding(.horizontal, 4)
        }
        .padding(16)
        .background(
            RoundedRectangle(cornerRadius: 24)
                .fill(Color(white: 0.08).opacity(0.6))
                .overlay(
                    RoundedRectangle(cornerRadius: 24)
                        .stroke(Color.white.opacity(0.06), lineWidth: 1)
                )
        )
    }
    
    private func findNearestPoint(x: CGFloat, width: CGFloat) -> StabilityPoint? {
        guard points.count > 0 else { return nil }
        guard let first = points.first else { return nil }
        let lastTime = max(1.0, points.last?.time ?? 1.0)
        let totalRange = lastTime - first.time
        guard totalRange > 0 else { return first }
        
        let fraction = Double(x / width)
        let targetTime = first.time + fraction * totalRange
        return points.min(by: { abs($0.time - targetTime) < abs($1.time - targetTime) })
    }
    
    // Grid Lines Drawer
    @ViewBuilder
    private func gridLines(width: CGFloat, height: CGFloat) -> some View {
        ZStack {
            ForEach([0.0, 0.25, 0.5, 0.75, 1.0], id: \.self) { fraction in
                let y = height * CGFloat(fraction)
                
                if fraction > 0.0 && fraction < 1.0 {
                    Path { path in
                        path.move(to: CGPoint(x: 0, y: y))
                        path.addLine(to: CGPoint(x: width, y: y))
                    }
                    .stroke(Color.white.opacity(0.04), lineWidth: 1)
                }
                
                let scoreText = "\(Int(100 - (fraction * 100)))%"
                Text(scoreText)
                    .font(.system(size: 8, weight: .medium, design: .monospaced))
                    .foregroundColor(.secondary.opacity(0.7))
                    .position(x: 18, y: y - 6)
            }
        }
    }
    
    private func getXPosition(for time: TimeInterval, width: CGFloat) -> CGFloat? {
        guard let first = points.first else { return nil }
        let lastTime = max(1.0, points.last?.time ?? 1.0)
        let totalRange = lastTime - first.time
        guard totalRange > 0 else { return 0 }
        
        let fraction = (time - first.time) / totalRange
        return CGFloat(fraction) * width
    }
    
    private func chartPath(in size: CGSize, isFilled: Bool) -> Path {
        var path = Path()
        guard points.count > 0 else { return path }
        
        let first = points.first!
        let lastTime = max(1.0, points.last?.time ?? 1.0)
        let totalRange = lastTime - first.time
        
        func getCGPoint(for pt: StabilityPoint) -> CGPoint {
            let xFraction = totalRange == 0 ? 0.0 : (pt.time - first.time) / totalRange
            let yFraction = pt.score / 100.0
            
            let x = CGFloat(xFraction) * size.width
            let y = size.height - CGFloat(yFraction) * size.height
            
            return CGPoint(
                x: min(size.width, max(0, x)),
                y: min(size.height, max(0, y))
            )
        }
        
        if isFilled {
            path.move(to: CGPoint(x: 0, y: size.height))
        }
        
        let firstPoint = getCGPoint(for: first)
        if isFilled {
            path.addLine(to: firstPoint)
        } else {
            path.move(to: firstPoint)
        }
        
        if points.count == 1 {
            path.addLine(to: CGPoint(x: size.width, y: firstPoint.y))
            if isFilled {
                path.addLine(to: CGPoint(x: size.width, y: size.height))
                path.closeSubpath()
            }
            return path
        }
        
        for idx in 0..<(points.count - 1) {
            let p0 = getCGPoint(for: points[idx])
            let p1 = getCGPoint(for: points[idx + 1])
            
            let control1 = CGPoint(x: p0.x + (p1.x - p0.x) / 2.0, y: p0.y)
            let control2 = CGPoint(x: p0.x + (p1.x - p0.x) / 2.0, y: p1.y)
            
            path.addCurve(to: p1, control1: control1, control2: control2)
        }
        
        if isFilled {
            let finalPoint = getCGPoint(for: points.last!)
            path.addLine(to: CGPoint(x: finalPoint.x, y: size.height))
            path.closeSubpath()
        }
        
        return path
    }
    
    private func formatTime(_ time: TimeInterval) -> String {
        let minutes = Int(time) / 60
        let seconds = Int(time) % 60
        return String(format: "%dm:%02ds", minutes, seconds)
    }
}

// ── Dual Stability Chart View for Client-Side Run Comparison ───────────────

struct DualStabilityChart: View {
    let pointsA: [StabilityPoint]
    let labelA: String
    let pointsB: [StabilityPoint]
    let labelB: String
    
    @State private var touchLocationX: CGFloat? = nil
    
    let colorA = Color(red: 0.0, green: 0.9, blue: 1.0)
    let colorB = Color(red: 1.0, green: 0.67, blue: 0.0)
    
    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            HStack {
                Text("📊 DUAL STABILITY COMPARISON")
                    .font(.system(size: 11, weight: .bold, design: .monospaced))
                    .foregroundColor(.secondary)
                Spacer()
                
                HStack(spacing: 12) {
                    HStack(spacing: 4) {
                        Circle().fill(colorA).frame(width: 8, height: 8)
                        Text(labelA)
                            .font(.system(size: 9, weight: .bold, design: .monospaced))
                            .foregroundColor(colorA)
                    }
                    HStack(spacing: 4) {
                        Circle().fill(colorB).frame(width: 8, height: 8)
                        Text(labelB)
                            .font(.system(size: 9, weight: .bold, design: .monospaced))
                            .foregroundColor(colorB)
                    }
                }
            }
            
            GeometryReader { geometry in
                let width = geometry.size.width
                let height = geometry.size.height
                
                ZStack(alignment: .topLeading) {
                    // Grid lines
                    ForEach([0.25, 0.5, 0.75], id: \.self) { fraction in
                        let y = height * CGFloat(fraction)
                        Path { path in
                            path.move(to: CGPoint(x: 0, y: y))
                            path.addLine(to: CGPoint(x: width, y: y))
                        }
                        .stroke(Color.white.opacity(0.04), lineWidth: 1)
                    }
                    
                    let maxTimeA = pointsA.last?.time ?? 1.0
                    let maxTimeB = pointsB.last?.time ?? 1.0
                    let totalMax = max(1.0, max(maxTimeA, maxTimeB))
                    
                    // Curve A
                    drawCurve(points: pointsA, totalTime: totalMax, color: colorA, in: geometry.size)
                    // Curve B
                    drawCurve(points: pointsB, totalTime: totalMax, color: colorB, in: geometry.size)
                    
                    // Selector line (Only visible while touching)
                    if let tx = touchLocationX {
                        Path { path in
                            path.move(to: CGPoint(x: tx, y: 0))
                            path.addLine(to: CGPoint(x: tx, y: height))
                        }
                        .stroke(Color.white.opacity(0.5), style: StrokeStyle(lineWidth: 1.5, dash: [6, 4]))
                    }
                }
                .contentShape(Rectangle())
                .gesture(
                    DragGesture(minimumDistance: 0)
                        .onChanged { value in
                            let locX = min(width, max(0, value.location.x))
                            touchLocationX = locX
                        }
                        .onEnded { _ in
                            touchLocationX = nil
                        }
                )
            }
            .frame(height: 200)
        }
        .padding(16)
        .background(
            RoundedRectangle(cornerRadius: 24)
                .fill(Color(white: 0.08).opacity(0.6))
                .overlay(
                    RoundedRectangle(cornerRadius: 24)
                        .stroke(Color.white.opacity(0.06), lineWidth: 1)
                )
        )
    }
    
    @ViewBuilder
    private func drawCurve(points: [StabilityPoint], totalTime: Double, color: Color, in size: CGSize) -> some View {
        if points.count > 0 {
            Path { path in
                let first = points.first!
                let getPoint: (StabilityPoint) -> CGPoint = { pt in
                    let x = CGFloat(pt.time / totalTime) * size.width
                    let y = size.height - CGFloat(pt.score / 100.0) * size.height
                    return CGPoint(x: min(size.width, max(0, x)), y: min(size.height, max(0, y)))
                }
                
                path.move(to: getPoint(first))
                for i in 0..<(points.count - 1) {
                    let p0 = getPoint(points[i])
                    let p1 = getPoint(points[i + 1])
                    let c1 = CGPoint(x: p0.x + (p1.x - p0.x) / 2.0, y: p0.y)
                    let c2 = CGPoint(x: p0.x + (p1.x - p0.x) / 2.0, y: p1.y)
                    path.addCurve(to: p1, control1: c1, control2: c2)
                }
            }
            .stroke(color, style: StrokeStyle(lineWidth: 3, lineCap: .round, lineJoin: .round))
        }
    }
}

File: ThermoHammerAndroid/app/src/main/java/com/example/thermohammer/ui/components/StabilityChart.kt
package com.example.thermohammer.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.*
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.layout.onGloballyPositioned
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.thermohammer.engine.StabilityPoint
import com.example.thermohammer.engine.ThermalEvent
import com.example.thermohammer.engine.ThermalState
import java.util.Locale

// ── Color Helpers ─────────────────────────────────────────────────────────────

fun thermalColor(state: ThermalState): Color = when (state) {
    ThermalState.NOMINAL  -> Color(0xFF33CC66)
    ThermalState.FAIR     -> Color(0xFFF2C94C)
    ThermalState.SERIOUS  -> Color(0xFFF2994A)
    ThermalState.CRITICAL -> Color(0xFFEB5757)
}

fun thermalName(state: ThermalState): String = when (state) {
    ThermalState.NOMINAL  -> "NOMINAL"
    ThermalState.FAIR     -> "FAIR"
    ThermalState.SERIOUS  -> "SERIOUS (THROTTLED)"
    ThermalState.CRITICAL -> "CRITICAL"
}

fun stabilityColor(score: Float): Color = when {
    score >= 90f -> Color(0xFF33CC66)
    score >= 75f -> Color(0xFFF2C94C)
    else         -> Color(0xFFE85520)
}

// ── Interactive Single Stability Chart ───────────────────────────────────────

@Composable
fun StabilityChart(
    points: List<StabilityPoint>,
    events: List<ThermalEvent>,
    modifier: Modifier = Modifier
) {
    var touchX by remember { mutableStateOf<Float?>(null) }
    var chartWidthPx by remember { mutableFloatStateOf(1f) }

    val selectedPoint = remember(touchX, points, chartWidthPx) {
        val tx = touchX ?: return@remember null
        if (points.isEmpty()) return@remember null
        val firstTime = points.first().time
        val lastTime = maxOf(1f, points.last().time)
        val totalRange = lastTime - firstTime
        val w = maxOf(1f, chartWidthPx)
        points.minByOrNull { pt ->
            val ptX = if (totalRange == 0f) 0f else (((pt.time - firstTime) / totalRange) * w)
            kotlin.math.abs(ptX - tx)
        }
    }

    Column(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(24.dp))
            .background(Color(0xFF141414))
            .border(1.dp, Color.White.copy(alpha = 0.06f), RoundedCornerShape(24.dp))
            .padding(16.dp)
    ) {
        // Header with Touch Point Info Display
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            Text(
                text = "▦  UNIFIED STABILITY CHART",
                style = TextStyle(
                    color = Color.White.copy(alpha = 0.5f),
                    fontSize = 11.sp,
                    fontFamily = FontFamily.Monospace,
                    fontWeight = FontWeight.Bold
                )
            )

            // Touch Stamp Badge Header (Disappears when touch is released)
            if (touchX != null && selectedPoint != null) {
                val pt = selectedPoint
                val tSec = pt.time.toInt()
                val minStr = tSec / 60
                val secStr = "%02d".format(tSec % 60)
                Row(
                    modifier = Modifier
                        .clip(RoundedCornerShape(8.dp))
                        .background(Color(0xFF00E5FF).copy(alpha = 0.15f))
                        .border(1.dp, Color(0xFF00E5FF).copy(alpha = 0.3f), RoundedCornerShape(8.dp))
                        .padding(horizontal = 8.dp, vertical = 4.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(6.dp)
                ) {
                    Text(
                        "${minStr}m:${secStr}s",
                        style = TextStyle(color = Color(0xFF00E5FF), fontSize = 10.sp, fontFamily = FontFamily.Monospace, fontWeight = FontWeight.Bold)
                    )
                    Text(
                        "SCORE: %.1f%%".format(Locale.US, pt.score),
                        style = TextStyle(color = stabilityColor(pt.score), fontSize = 10.sp, fontFamily = FontFamily.Monospace, fontWeight = FontWeight.Black)
                    )
                }
            }
        }
        Spacer(Modifier.height(12.dp))

        // Chart Canvas with Touch Drag / Release Gestures
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(180.dp)
                .onGloballyPositioned { chartWidthPx = it.size.width.toFloat() }
        ) {
            Canvas(
                modifier = Modifier
                    .fillMaxSize()
                    .pointerInput(points) {
                        detectDragGestures(
                            onDragStart = { offset -> touchX = offset.x },
                            onDrag = { change, _ -> touchX = change.position.x },
                            onDragEnd = { touchX = null },
                            onDragCancel = { touchX = null }
                        )
                    }
                    .pointerInput(points) {
                        detectTapGestures(
                            onPress = { offset ->
                                touchX = offset.x
                                tryAwaitRelease()
                                touchX = null
                            }
                        )
                    }
            ) {
                val w = size.width
                val h = size.height
                chartWidthPx = w

                // Grid lines at 25%, 50%, 75%
                listOf(0.25f, 0.50f, 0.75f).forEach { fraction ->
                    val y = h * fraction
                    drawLine(
                        color = Color.White.copy(alpha = 0.04f),
                        start = Offset(0f, y),
                        end = Offset(w, y),
                        strokeWidth = 1f
                    )
                }

                if (points.isEmpty()) return@Canvas

                val firstTime = points.first().time
                val lastTime = maxOf(1f, points.last().time)
                val totalRange = lastTime - firstTime

                fun getX(time: Float): Float {
                    if (totalRange == 0f) return 0f
                    return ((time - firstTime) / totalRange) * w
                }
                fun getY(score: Float): Float = h - (score / 100f) * h

                // Line gradient
                val lineGradient = Brush.verticalGradient(
                    colorStops = arrayOf(
                        0.0f to Color(0xFF33CC66),
                        0.25f to Color(0xFFF2C94C),
                        0.5f to Color(0xFFE85520),
                        1.0f to Color(0xFFD93025)
                    ),
                    startY = 0f, endY = h
                )

                // Fill gradient
                val fillGradient = Brush.verticalGradient(
                    colorStops = arrayOf(
                        0.0f to Color(0xFF33CC66).copy(alpha = 0.15f),
                        0.5f to Color(0xFFE85520).copy(alpha = 0.06f),
                        1.0f to Color.Transparent
                    ),
                    startY = 0f, endY = h
                )

                val path = Path()
                val fillPath = Path()

                fillPath.moveTo(getX(points.first().time), h)
                fillPath.lineTo(getX(points.first().time), getY(points.first().score))

                if (points.size == 1) {
                    path.moveTo(0f, getY(points[0].score))
                    path.lineTo(w, getY(points[0].score))
                    fillPath.lineTo(w, getY(points[0].score))
                    fillPath.lineTo(w, h)
                    fillPath.close()
                } else {
                    path.moveTo(getX(points[0].time), getY(points[0].score))
                    for (i in 0 until points.size - 1) {
                        val x0 = getX(points[i].time); val y0 = getY(points[i].score)
                        val x1 = getX(points[i+1].time); val y1 = getY(points[i+1].score)
                        val cx = x0 + (x1 - x0) / 2f
                        path.cubicTo(cx, y0, cx, y1, x1, y1)
                        fillPath.cubicTo(cx, y0, cx, y1, x1, y1)
                    }
                    val last = points.last()
                    fillPath.lineTo(getX(last.time), h)
                    fillPath.close()
                }

                drawPath(fillPath, fillGradient)
                drawPath(path, lineGradient, style = Stroke(width = 3.5f, cap = StrokeCap.Round, join = StrokeJoin.Round))

                // Thermal event markers
                for (event in events) {
                    val xPos = getX(event.time)
                    val color = thermalColor(event.state)
                    val dotLen = 6f; val gap = 4f
                    var y = 24f
                    while (y < h) {
                        drawLine(color.copy(alpha = 0.5f), Offset(xPos, y), Offset(xPos, minOf(y + dotLen, h)), 1f)
                        y += dotLen + gap
                    }
                }

                // Touch Inspector Cursor Line and Highlight Node (Only drawn while touching)
                touchX?.let { tx ->
                    val clampedX = tx.coerceIn(0f, w)
                    val nearest = points.minByOrNull { kotlin.math.abs(getX(it.time) - clampedX) }
                    nearest?.let { targetPt ->
                        val targetX = getX(targetPt.time)
                        val targetY = getY(targetPt.score)

                        // Vertical selector line
                        drawLine(
                            color = Color(0xFF00E5FF).copy(alpha = 0.7f),
                            start = Offset(targetX, 0f),
                            end = Offset(targetX, h),
                            strokeWidth = 1.5f,
                            pathEffect = PathEffect.dashPathEffect(floatArrayOf(8f, 6f), 0f)
                        )

                        // Glowing Node Outer Aura
                        drawCircle(
                            color = Color(0xFF00E5FF).copy(alpha = 0.25f),
                            radius = 12f,
                            center = Offset(targetX, targetY)
                        )

                        // Glowing Node Inner Solid Circle
                        drawCircle(
                            color = Color(0xFF00E5FF),
                            radius = 5f,
                            center = Offset(targetX, targetY)
                        )
                        drawCircle(
                            color = Color.White,
                            radius = 2.5f,
                            center = Offset(targetX, targetY)
                        )
                    }
                }
            }
        }

        // X-axis labels
        Spacer(Modifier.height(8.dp))
        Row(Modifier.fillMaxWidth()) {
            Text("0m:00s", style = TextStyle(color = Color.White.copy(alpha = 0.4f), fontSize = 9.sp, fontFamily = FontFamily.Monospace))
            Spacer(Modifier.weight(1f))
            val lastTime = points.lastOrNull()?.time?.toInt() ?: 0
            Text(
                "${lastTime / 60}m:${"%02d".format(lastTime % 60)}s",
                style = TextStyle(color = Color.White.copy(alpha = 0.4f), fontSize = 9.sp, fontFamily = FontFamily.Monospace)
            )
        }
    }
}

// ── Dual Stability Chart for Client-Side Run Comparison ─────────────────────

@Composable
fun DualStabilityChart(
    pointsA: List<StabilityPoint>,
    labelA: String,
    pointsB: List<StabilityPoint>,
    labelB: String,
    modifier: Modifier = Modifier
) {
    var touchX by remember { mutableStateOf<Float?>(null) }
    var selectedA by remember { mutableStateOf<StabilityPoint?>(null) }
    var selectedB by remember { mutableStateOf<StabilityPoint?>(null) }

    val colorA = Color(0xFF00E5FF) // Cyan
    val colorB = Color(0xFFFFAB00) // Amber

    Column(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(24.dp))
            .background(Color(0xFF141414))
            .border(1.dp, Color.White.copy(alpha = 0.08f), RoundedCornerShape(24.dp))
            .padding(16.dp)
    ) {
        // Legend & Header
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            Text(
                text = "📊  DUAL STABILITY COMPARISON",
                style = TextStyle(color = Color.White.copy(alpha = 0.6f), fontSize = 11.sp, fontFamily = FontFamily.Monospace, fontWeight = FontWeight.Bold)
            )

            Row(horizontalArrangement = Arrangement.spacedBy(12.dp), verticalAlignment = Alignment.CenterVertically) {
                Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(4.dp)) {
                    Box(Modifier.size(8.dp).clip(RoundedCornerShape(2.dp)).background(colorA))
                    Text(labelA, style = TextStyle(color = colorA, fontSize = 9.sp, fontFamily = FontFamily.Monospace, fontWeight = FontWeight.Bold))
                }
                Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(4.dp)) {
                    Box(Modifier.size(8.dp).clip(RoundedCornerShape(2.dp)).background(colorB))
                    Text(labelB, style = TextStyle(color = colorB, fontSize = 9.sp, fontFamily = FontFamily.Monospace, fontWeight = FontWeight.Bold))
                }
            }
        }

        Spacer(Modifier.height(12.dp))

        // Canvas overlay
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(200.dp)
        ) {
            Canvas(
                modifier = Modifier
                    .fillMaxSize()
                    .pointerInput(pointsA, pointsB) {
                        detectDragGestures(
                            onDragStart = { offset -> touchX = offset.x },
                            onDrag = { change, _ -> touchX = change.position.x },
                            onDragEnd = { touchX = null; selectedA = null; selectedB = null },
                            onDragCancel = { touchX = null; selectedA = null; selectedB = null }
                        )
                    }
                    .pointerInput(pointsA, pointsB) {
                        detectTapGestures(
                            onPress = { offset ->
                                touchX = offset.x
                                tryAwaitRelease()
                                touchX = null; selectedA = null; selectedB = null
                            }
                        )
                    }
            ) {
                val w = size.width
                val h = size.height

                // Grid lines
                listOf(0.25f, 0.50f, 0.75f).forEach { fraction ->
                    val y = h * fraction
                    drawLine(color = Color.White.copy(alpha = 0.04f), start = Offset(0f, y), end = Offset(w, y), strokeWidth = 1f)
                }

                val maxTimeA = pointsA.lastOrNull()?.time ?: 1f
                val maxTimeB = pointsB.lastOrNull()?.time ?: 1f
                val totalMaxTime = maxOf(1f, maxOf(maxTimeA, maxTimeB))

                fun getX(time: Float): Float = (time / totalMaxTime) * w
                fun getY(score: Float): Float = h - (score / 100f) * h

                // Helper to draw path
                fun drawCurve(pts: List<StabilityPoint>, color: Color) {
                    if (pts.isEmpty()) return
                    val path = Path()
                    path.moveTo(getX(pts[0].time), getY(pts[0].score))
                    for (i in 0 until pts.size - 1) {
                        val x0 = getX(pts[i].time); val y0 = getY(pts[i].score)
                        val x1 = getX(pts[i+1].time); val y1 = getY(pts[i+1].score)
                        val cx = x0 + (x1 - x0) / 2f
                        path.cubicTo(cx, y0, cx, y1, x1, y1)
                    }
                    drawPath(path, color, style = Stroke(width = 3f, cap = StrokeCap.Round, join = StrokeJoin.Round))
                }

                // Draw curves
                drawCurve(pointsA, colorA)
                drawCurve(pointsB, colorB)

                // Interactive touch vertical line & data nodes (Resets when touch released)
                touchX?.let { tx ->
                    val clampedX = tx.coerceIn(0f, w)
                    val targetA = pointsA.minByOrNull { kotlin.math.abs(getX(it.time) - clampedX) }
                    val targetB = pointsB.minByOrNull { kotlin.math.abs(getX(it.time) - clampedX) }

                    selectedA = targetA
                    selectedB = targetB

                    // Vertical dashed indicator
                    drawLine(
                        color = Color.White.copy(alpha = 0.5f),
                        start = Offset(clampedX, 0f),
                        end = Offset(clampedX, h),
                        strokeWidth = 1.5f,
                        pathEffect = PathEffect.dashPathEffect(floatArrayOf(6f, 6f), 0f)
                    )

                    targetA?.let { pA ->
                        val xA = getX(pA.time); val yA = getY(pA.score)
                        drawCircle(colorA, radius = 5f, center = Offset(xA, yA))
                        drawCircle(Color.White, radius = 2.5f, center = Offset(xA, yA))
                    }

                    targetB?.let { pB ->
                        val xB = getX(pB.time); val yB = getY(pB.score)
                        drawCircle(colorB, radius = 5f, center = Offset(xB, yB))
                        drawCircle(Color.White, radius = 2.5f, center = Offset(xB, yB))
                    }
                }
            }
        }

        // Live Selected Data Bar (Only visible while touching)
        if (touchX != null && (selectedA != null || selectedB != null)) {
            Spacer(Modifier.height(8.dp))
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .clip(RoundedCornerShape(12.dp))
                    .background(Color.White.copy(alpha = 0.04f))
                    .padding(horizontal = 12.dp, vertical = 8.dp),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                selectedA?.let { pA ->
                    val sec = pA.time.toInt()
                    Text(
                        "$labelA: %02d:%02d -> %.1f%%".format(sec / 60, sec % 60, pA.score),
                        style = TextStyle(color = colorA, fontSize = 10.sp, fontFamily = FontFamily.Monospace, fontWeight = FontWeight.Bold)
                    )
                }
                selectedB?.let { pB ->
                    val sec = pB.time.toInt()
                    Text(
                        "$labelB: %02d:%02d -> %.1f%%".format(sec / 60, sec % 60, pB.score),
                        style = TextStyle(color = colorB, fontSize = 10.sp, fontFamily = FontFamily.Monospace, fontWeight = FontWeight.Bold)
                    )
                }
            }
        }
    }
}
